src/gt4py/next/embedded/function_field.py

```diff
@@ -0,0 +1,203 @@
+# GT4Py - GridTools Framework
+#
+# Copyright (c) 2014-2023, ETH Zurich
+# All rights reserved.
+#
+# This file is part of the GT4Py project and the GridTools framework.
+# GT4Py is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or any later
+# version. See the LICENSE.txt file at the top-level directory of this
+# distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+from __future__ import annotations
+
+import dataclasses
+import operator
+from typing import Any, Callable, TypeAlias, TypeGuard
+
+import numpy as np
+
+from gt4py._core import definitions as core_defs
+from gt4py.next import common
+from gt4py.next.common import Infinity
+from gt4py.next.embedded import common as embedded_common, nd_array_field as nd
+from gt4py.next.embedded.nd_array_field import _get_slices_from_domain_slice
+
+
+_EMPTY_DOMAIN = common.Domain((), ())
+
+ConstantFieldValue: TypeAlias = int | float | complex
+
+
+@dataclasses.dataclass(frozen=True)
+class FunctionField:
+    func: Callable
+    domain: common.Domain = _EMPTY_DOMAIN
+    _constant: bool = False
+
+    def restrict(self, index: common.FieldSlice) -> FunctionField:
+        if _has_empty_domain(self):
+            raise IndexError("Cannot slice ConstantField without a Domain.")
+        new_domain = embedded_common.sub_domain(self.domain, index)
+        return self.__class__(self.func, new_domain)
+
+    __getitem__ = restrict
+
+    @property
+    def ndarray(self) -> core_defs.NDArrayObject:
+        if _has_empty_domain(self):
+            raise ValueError("Cannot get ndarray for FunctionField without Domain.")
+
+        shape = []
+        for _, rng in self.domain:
+            if Infinity.positive() in (abs(rng.start), abs(rng.stop)):
+                raise ValueError(
+                    f"Cannot construct ndarray with infinite range in Domain: {self.domain}"
+                )
+            else:
+                shape.append(len(rng))
+
+        if self._constant:
+            return np.full(shape, self.func())
+
+        return np.fromfunction(lambda *indices: self.func(*indices), shape)
+
+    def _handle_identity_op(self, other: FunctionField, operator_func: Callable) -> FunctionField:
```

Comment:
```suggestion
    def _handle_function_field_op(self, other: FunctionField, operator_func: Callable) -> FunctionField:
```
not perfect, but `identity_op` pretty much has the meaning of `lambda x: x`